examples/fluids/src/dirichlet.c

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2017-2022, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+#include "../navierstokes.h"
+#include "../qfunctions/dirichlet_boundary.h"
+#include "../problems/stg_shur14.h"
+
+PetscErrorCode SetupStrongSTG_Ceed(Ceed ceed, CeedData ceed_data, DM dm,
+                                   AppCtx app_ctx, ProblemData *problem, SimpleBC bc,
+                                   CeedInt Q_sur, CeedInt q_data_size_sur, CeedOperator op_dirichlet) {
+  CeedInt              num_comp_x=problem->dim, num_comp_q = 5, num_elem,
+                       elem_size;
+  CeedVector           multiplicity, x_stored, scale_stored, q_data_sur;
+  CeedBasis            basis_x_to_q_sur;
+  CeedElemRestriction  elem_restr_x_sur, elem_restr_q_sur, elem_restr_x_stored,
+                       elem_restr_scale, elem_restr_qd_sur;
+  CeedQFunction        qf_setup, qf_strongbc;
+  CeedOperator         op_setup, op_dirichlet_sub, op_setup_sur;
+  PetscFunctionBeginUser;
+
+  DMLabel domain_label;
+  PetscCall(DMGetLabel(dm, "Face Sets", &domain_label));
+
+  // Basis
+  CeedInt height = 1;
+  PetscCall(CeedBasisCreateProjection(ceed_data->basis_x_sur,
+                                      ceed_data->basis_q_sur, &basis_x_to_q_sur));
+
+  // Setup QFunction
+  CeedQFunctionCreateInterior(ceed, 1, SetupDirichletBC, SetupDirichletBC_loc,
+                              &qf_setup);
+  CeedQFunctionAddInput(qf_setup, "x", num_comp_x, CEED_EVAL_INTERP);
+  CeedQFunctionAddInput(qf_setup, "multiplicity", num_comp_q, CEED_EVAL_NONE);
+  CeedQFunctionAddOutput(qf_setup, "x stored", num_comp_x, CEED_EVAL_NONE);
+  CeedQFunctionAddOutput(qf_setup, "scale", 1, CEED_EVAL_NONE);
+
+  // Compute contribution on each boundary face
+  for (CeedInt i=0; i < bc->num_inflow; i++) {
+    // -- Restrictions
+    PetscCall(GetRestrictionForDomain(ceed, dm, height, domain_label,
```

Comment:
Here, this Restriction, I think, has 5 components, so that 5th component is dumping junk into the local vector for some of the GPU backends.

Reply:
Yep, that should be fixed now, but it's still failing with `/gpu/hip/gen` on noether.

Reply:
It's still failing, even with the latest commit (which forces the unset output to be zeroed out).

Reply:
I looked over the code a few times and don't see any obvious reason why this would be the case. It might be useful to check the l-vec after boundary value insertion on a small problem with `/gpu/hip/shared` and `/gpu/hip/gen`

It is probably too much information, but you can see the generated code by setting the environment variable `CEED_DEBUG=1` or `DEBUG=1`

Reply:
Oh, is this `/gpu/hip/gen` issue still unresolved?